docs/extensions/server/applications.rst

```diff
@@ -113,4 +82,54 @@ The steps of adding a custom Application Operation
           url: Option[String] = None,
           error: Option[String] = None)
 
+For application state mapping, you can reference the implementation of yarn:
+
+   .. code-block:: scala
+
+      def toApplicationState(state: YarnApplicationState): ApplicationState = state match {
+        case YarnApplicationState.NEW => ApplicationState.PENDING
+        case YarnApplicationState.NEW_SAVING => ApplicationState.PENDING
+        case YarnApplicationState.SUBMITTED => ApplicationState.PENDING
+        case YarnApplicationState.ACCEPTED => ApplicationState.PENDING
+        case YarnApplicationState.RUNNING => ApplicationState.RUNNING
+        case YarnApplicationState.FINISHED => ApplicationState.FINISHED
+        case YarnApplicationState.FAILED => ApplicationState.FAILED
+        case YarnApplicationState.KILLED => ApplicationState.KILLED
+        case _ =>
+          warn(s"The yarn driver state: $state is not supported, " +
+            "mark the application state as UNKNOWN.")
+          ApplicationState.UNKNOWN
+      }
+
+Build A Custom Application Operation
+------------------------------------
+
+- reference kyuubi-server
+
+   .. code-block:: xml
+
+      <dependency>
+         <groupId>org.apache.kyuubi</groupId>
+         <artifactId>kyuubi-server_2.12</artifactId>
+         <version>1.5.2-incubating</version>
+         <scope>provided</scope>
+      </dependency>
+
+- create a custom class which implements the ``org.apache.kyuubi.engine.ApplicationOperation``.
+
+Enable Custom Application Operation
+-----------------------------------
+
+.. note:: Kyuubi uses Java SPI to load the custom Application Operation
+
+1. create a directory META-INF.services and a file with the fully-qualified name of the interface ``ApplicationOperation``:
```

Comment:
this is better to be added in the steps of building tar ball?

Reply:
do you mean add `mvn package` in this stpes ?

Reply:
no, I mean this belongs to the above section

Reply:
got it